### ISSUE TITLE
Uncaught RangeError: Maximum call stack size exceeded

### ISSUE THREAD

**Issue by mwweb** [vdm-io/Joomla-Component-Builder#146]
I'm still doing testing with 2.5.8, since I wanted to ensure that everything is working fine.

In JCB 2.5.5 everything would compile fine, and everything would load properly for my component on the frontend and the backend.

After the upgrade to JCB 2.5.8 I was doing some testing and was getting Error 0, which is such a non-descriptive error message. In looking at the call stack, it looks like under JCB 2.5.8 it was having issues with the joined com_categories table:

```

1 | () | JROOT/libraries/joomla/database/driver/mysqli.php:660
-- | -- | --
2 | JDatabaseDriverMysqli->execute() | JROOT/components/com_mostwantedrealestate/models/properties.php:324
3 | MostwantedrealestateModelProperties->getCatidIdCeef_B() | JROOT/components/com_mostwantedrealestate/models/properties.php:249

```

In looking at the dynamic get, the joined field is there, but I tested deleting then re-adding the table, which is when I noticed that it wasn't populating the selection area. Some other views were loading.

Digging deeper, there is a jQuery issue present on the admin views. I mentioned this during the staging, but it looks like it is still an issue. This issue is present on localhost, and if I backup and install the dev on my web hosting server.
```
Uncaught RangeError: Maximum call stack size exceeded
    at Function.Sizzle (jquery.js?31617fba9b2d24f9efd1ac0ebd420dd2:844)
    at Function.jQuery.find (jquery-migrate.js?31617fba9b2d24f9efd1ac0ebd420dd2:290)
    at jQuery.fn.init.find (jquery.js?31617fba9b2d24f9efd1ac0ebd420dd2:2826)
    at jQuery.fn.init.jQuery.fn.find (jquery-migrate.js?31617fba9b2d24f9efd1ac0ebd420dd2:677)
    at jQuery.fn.init (jquery.js?31617fba9b2d24f9efd1ac0ebd420dd2:2949)
    at new jQuery.fn.init (jquery-migrate.js?31617fba9b2d24f9efd1ac0ebd420dd2:241)
    at jQuery (jquery.js?31617fba9b2d24f9efd1ac0ebd420dd2:75)
    at HTMLDocument.<anonymous> (index.php?option=com_componentbuilder&view=dynamic_get&layout=edit&id=92:196)
    at fire (jquery.js?31617fba9b2d24f9efd1ac0ebd420dd2:3232)
    at Object.fireWith [as resolveWith] (jquery.js?31617fba9b2d24f9efd1ac0ebd420dd2:3362)
```

**Comment by Llewellynvdm** [comment]
I think when looking at your code dump, that you have a cache issue, please make sure to wipe your browser history completely and test again.

**Comment by Llewellynvdm** [comment]
Just a side note, JCB upgrade did not change the compiler in anyway. They way the compiler worked before, so it still works. In fact the data model of the subforms are identical to how the data model use to look, we use to convert it to that data model, so the subfrom array structure is in fact unchanged.

Compiling huge very complex components with JCB have been tested in git repositories and the result is only the change of date and version... so the compilation is still delivering the same result.

**Comment by mwweb** [comment]
I routinely clear my cache, but to test I logged out of everything, cleared cache in chrome "from the beginning of time". Closed my browser, reopened. Logged back into Joomla. Cleared site and admin cache. Cleared expired cache.

I then went back into the dynamic get and did a CTRL F5, and still persists:

![jcb_jquery_error](https://user-images.githubusercontent.com/19194012/31966287-11b44d40-b8bf-11e7-8da5-05dc590834d1.png)

**Comment by Llewellynvdm** [comment]
Clearing your cache is not enough, you must clear your browser history.

**Comment by Llewellynvdm** [comment]
Well lets putt it this way... how many browsers do you have on your computer, if they all are giving the same error then we have reason to believe that we have a bug... but if only the one you always use do that, we would tend to thing it is the browser and not the program.

**Comment by Llewellynvdm** [comment]
If it as a bug that only happens in one browser, as it is possible. Then we should see that bug on our computers in that same browser... and the fact is I don't see that bug in chrome on my PC.

**Comment by Llewellynvdm** [comment]
The code, or error response that you are sending me does not help really. There is not even any reason to believe that it is related to JCB... not that I am saying that it isn't, just that it does not give me anything to work with.

**Comment by Llewellynvdm** [comment]
Some browsers store the JavaScript in the browser memory, to speedup the page load. When changes are made to the JavaScript these changes conflict with what is in memory. I cleared my memory and now all works well... that is why I think it is browser memory issue. I can be wrong... so let me know what you have experienced.

**Comment by Llewellynvdm** [comment]
![image](https://user-images.githubusercontent.com/5607939/31967073-0c33f69c-b90d-11e7-9404-5ac36000edc9.png)
Here you can see me load the category table values (in chrome), and they all loaded with no errors.

**Comment by Llewellynvdm** [comment]
I have another question, would you mind to share to code around line 249 in JROOT/components/com_mostwantedrealestate/models/properties.php

So I can look at the compiled code... then I will also need a few screen shots of the dynamic get that is used to build that model.

I think there are two issues here:
- The dynamicGet view... and its JavaScript that loads the table values into the selection area.
- The compiled model based on the dynamic get

My first response is only targeting the JavaScript issue. Lets see what we can see around the compiled issue.

**Comment by mwweb** [comment]
It appears to only be affecting Chrome. If I go to my dev site in Firefox, or Microsoft Edge, it works fine.

Clearing everything (except saved passwords) in Chrome isn't resolving it. I don't have the time to deal with Chrome acting up (whole day wasted due to Chrome apparently), so I'm going to have to dump Chrome for dev, and switch to Firefox or Edge.

**Comment by Llewellynvdm** [comment]
@mwweb if you switch use Firefox and only pull on the other in testing.

You should check out https://www.chromium.org/ 
The Chromium projects include Chromium and Chromium OS, the open-source projects behind the Google Chrome browser and Google Chrome OS, respectively.

You can have both installed at once... well on Ubuntu anyway. So I run lets see...
- Firefox
- Chrome
- Chromium
- Opera
- Konqueror
- Iceweasel
- Lynx (works only in commandline)

Just to name the once I use often... testing and testing and testing... sometimes feels like madness.

Hey, my wife gave me a great encouragement the other day with a slogan she found on-line, not sure where. It said:
>Genius is perseverance in disguise

**Comment by mwweb** [comment]
Here's the actual complete compiled model. This is one, but there are 4 or 5 different models that do the same thing, only since 2.5.8. My next step is going to be to cross compare a 2.5.5 compile to a 2.5.8 compile.

[properties.zip](https://github.com/vdm-io/Joomla-Component-Builder/files/1413000/properties.zip)

Screenshots of the get

![dynamicget-1](https://user-images.githubusercontent.com/19194012/31976436-6e1a0f5c-b8ec-11e7-8a7e-80d822830652.png)
![dynamicget-2](https://user-images.githubusercontent.com/19194012/31976437-6e378294-b8ec-11e7-876a-cc5747631eac.png)
![dynamicget-3](https://user-images.githubusercontent.com/19194012/31976438-6e576bea-b8ec-11e7-82b6-c7d901a26901.png)
![dynamicget-4](https://user-images.githubusercontent.com/19194012/31976435-6ddd1cfa-b8ec-11e7-93b3-cf99e4ccb51b.png)

**Comment by mwweb** [comment]
OK. There is a difference in the compile at line 248-249:

2.5.5
`				// [Interpretation 1680] set catidIdCategoriesB to the $item object.
				$item->catidIdCategoriesB = $this->getCatidIdCategoriesCeef_B($item->catid);
`

2.5.8:
`				// [Interpretation 1680] set catidIdB to the $item object.
				$item->catidIdB = $this->getCatidIdCeef_B($item->catid);
`

Also, line 301-320.

2.5.5:

`	* Method to get an array of Categories Objects.
	*
	* @return mixed  An array of Categories Objects on success, false on failure.
	*
	*/
	public function getCatidIdCategoriesCeef_B($catid)
	{
		// [Interpretation 2445] Get a db connection.
		$db = JFactory::getDbo();

		// [Interpretation 2447] Create a new query object.
		$query = $db->getQuery(true);

		// [Interpretation 2449] Get from #__categories as b

		$query->select($db->quoteName(
			array('b.title'),
			array('title')));
		$query->from($db->quoteName('#__categories', 'b'));
		$query->where('b.id = ' . $db->quote($catid));
`

2.5.8:

`	* Method to get an array of  Objects.
	*
	* @return mixed  An array of  Objects on success, false on failure.
	*
	*/
	public function getCatidIdCeef_B($catid)
	{
		// [Interpretation 2445] Get a db connection.
		$db = JFactory::getDbo();

		// [Interpretation 2447] Create a new query object.
		$query = $db->getQuery(true);

		// [Interpretation 2449] Get from  as b
		



		$query->from();
		$query->where('b.id = ' . $db->quote($catid));
`

**Comment by Llewellynvdm** [comment]
Hmmm, okay thanks for sharing this... I will get on that and see if I can see the problem on my side. 

It is clearly dropping some code... wow, not sure how I missed that. I will get back to you!

**Comment by Llewellynvdm** [comment]
The Method that was changed that builds the dynamic get structure in the compiler is on [line 2134 ](https://github.com/vdm-io/Joomla-Component-Builder/blob/master/admin/helpers/compiler/a_Get.php#L2134) of the a_Get.php

This changed when we moved to v2.5.6 check the old method that [still used the repeatable field structure here](https://github.com/vdm-io/Joomla-Component-Builder/blob/v2.5.5/admin/helpers/compiler/a_Get.php#L2051) on line 2051 in v2.5.5

Let me know if you see a typo... I will also go over it a few times again. I know that is that only place where this bug should be... or so I think.

**Comment by mwweb** [comment]
It looks like there are some definite differences between the two compiler files. There would be some, of course, but some in that particular area of the code.

**Comment by Llewellynvdm** [comment]
Yes it has changed, that is what I said
> This changed when we moved to v2.5.6

I am asking if you are able to ready the code and see any typos or mismatching values. You can do `var_dump();exit;` at any of those lines and when the compiler runs you can look at the values.

I need a new set of eyes on this because to me it looks correct.

**Comment by mwweb** [comment]
I guess I'm not exactly understanding that you would like me to do to debug this. But I'll attempt to "fool around" with it and see if I can figure something out.

**Comment by ro-ot** [comment]
@mwweb can you check the json stored values of that Joined->db->Table 

I have been testing this, and I can't duplicate that same behavior. I used Sermon Distributor to test, and JCB builds all the models correctly as before. What happens to the category.php (model) around [line 274](https://github.com/SermonDistributor/Joomla-3-Component/blob/master/site/models/category.php#L277)

**Comment by mwweb** [comment]
I running some tests now, but this is for ro-ot.

If you want to check out what it is doing, please do the following:

1. Create a new get, let's call it "All Sermons".  It's a getListQuery. Source is Back-end View, View Talbe is Sermon, leave the selection as is.
2. On the Joint tab for "Join DB Tables" add the categories table, return row type as multiple as b type join outer on a.catid equal b.id. Selection I just reduced to b.title as title.
3. Save & Close
4. Go to Site View and create a really simple site view (I called it allsermons) with that new get as the Main Get.
5. Add site view to the component, then compile.

In the newly compiled site/model/allsermon.php here is the compiled output:

`	/**
	* Method to get an array of  Objects.
	*
	* @return mixed  An array of  Objects on success, false on failure.
	*
	*/
	public function getCatidIdEffb_B($catid)
	{
		// Get a db connection.
		$db = JFactory::getDbo();

		// Create a new query object.
		$query = $db->getQuery(true);

		// Get from  as b
		
		$query->from();
		$query->where('b.id = ' . $db->quote($catid));

		// Reset the query using our newly populated query object.
		$db->setQuery($query);
		$db->execute();

		// check if there was data returned
		if ($db->getNumRows())
		{
			return $db->loadObjectList();
		}
		return false;
	}
`

**Comment by ro-ot** [comment]
Just a head-up if you add code block use **three backtick** also called **grave accent**

If you use only one, it is for inline code. Three is for code blocks.

**Comment by mwweb** [comment]
The var_dump appears to be outputting the results current to screen, except an occasional undefined variable, but it looks like that may be due to existing the code prematurely.

'''Notice: Undefined variable: value in /var/www/dev/administrator/components/com_componentbuilder/helpers/compiler/a_Get.php on line 2310
array(9) { ["db_table"]=> string(10) "categories" ["row_type"]=> string(1) "1" ["as"]=> string(1) "b" ["type"]=> string(4) "LEFT" ["on_field"]=> string(7) "a.catid" ["operator"]=> string(1) "=" ["join_field"]=> string(4) "b.id" ["selection"]=> bool(false) ["key"]=> string(39) "preacher_sermons_preacher_id_twenty_two" } '''

**Comment by ro-ot** [comment]
Ok I got the same result this time, thanks @mwweb!

I also found the bug! the work around until the next release is to update a line in the compiler (a_Get.php) file. 

Was a very stupid mistake. I think we missed it because the method is so big, but yes we have an undeclared variable in the method called `$value` on line [2310](https://github.com/vdm-io/Joomla-Component-Builder/blob/master/admin/helpers/compiler/a_Get.php#L2310) in a_Get.php change it to `$option1['selection']` and problem solved.

So @Llewellynvdm was right there was a typo in that method!

**Comment by Llewellynvdm** [comment]
Great work, I will add this to the next update.

**Comment by mwweb** [comment]
Even in Firefox and MS Edge I still get the error popping up in the functionality, just doesn't show in the console output. I noticed it today when attempting to update a dynamic get. The selection area was not updating with the fields. So, I opened in Chrome, and there was the pesky error again.

In some reading, so far, what I have found is that the most common cause of "Maximum call stack size exceeded" most likely refers to one JavaScript function referring to another JavaScript function which returns back to the first function and therefor creates an endless loop.
